### PR TITLE
Add support for virtio-watchdog

### PR DESCRIPTION
**Pull request by rbradford**
Add a simple virtio-watchdog implementation. Need to merge change into our kernel fork before merging this PR.

Fixes: #1750

### CODE REVIEW

File: vm-virtio/src/lib.rs
Comment (sboeuf) [inline vm-virtio/src/lib.rs:65]: There's a mismatch here. Should be `35`.

File: virtio-devices/src/watchdog.rs
Comment (sboeuf) [inline virtio-devices/src/watchdog.rs:3]: Did you actually port this code from crosvm? Otherwise you should at least add a portion of copyright for Intel.

File: tests/integration.rs
Comment (sboeuf) [inline tests/integration.rs:5309]: Does it need to be x86 specific?

File: tests/integration.rs
Comment (sboeuf) [inline tests/integration.rs:5417]: That is truly a great integration test :+1:

File: virtio-devices/src/watchdog.rs
Comment (sboeuf) [inline virtio-devices/src/watchdog.rs:29]: I'm fine with 8, but I think 1 could work as well, right?

File: virtio-devices/src/watchdog.rs
Comment (sboeuf) [inline virtio-devices/src/watchdog.rs:39]: I think we could do better and be more flexible since the timer interval should be changeable from guest userspace, no matter what has been set by default in the driver. That's why I think we could introduce a `virtio_watchdog_config`, holding onto the timeout value. This value can be written from the driver so that we would catch a change in the config from the device.
And from the driver, that means we must implement `.set_timeout()` callback to be notified when the timeout is updated from `watchdog_set_timeout()`.

**Review by sboeuf — COMMENTED** [review]
Looks good!

File: vm-virtio/src/lib.rs
Comment (rbradford) [reply vm-virtio/src/lib.rs]: Fixed.

File: virtio-devices/src/watchdog.rs
Comment (rbradford) [reply virtio-devices/src/watchdog.rs:42]: I thought about it but that adds quite a lot more complexity. And AFAICT, when I was playing with this, systemd just wants to use 15s.

Comment (sboeuf) [reply virtio-devices/src/watchdog.rs:42]: That's fine, but that might be something VIRTIO maintainers will ask for. Anyway, we can still add this later if needed with a follow up PR.

File: tests/integration.rs
Comment (rbradford) [reply tests/integration.rs]: I tried enabling it for arm64 but the first reboot (nothing to do with the watchdog) fails.

Comment (rbradford) [reply tests/integration.rs:5309]: I will open a bug after this is merged to try and investigate this issue.

**Review by sboeuf — COMMENTED** [review]
LGTM given that you will remove the commit `HACK: Use temporary branch for virtio-watchdog support`.

**Comment by rbradford** [comment]
> LGTM given that you will remove the commit `HACK: Use temporary branch for virtio-watchdog support`.
Yes, I will merge cloud-hypervisor/linux#3 first but before I do that I will make a formal request for the ID.

File: virtio-devices/src/watchdog.rs
Comment (sameo) [inline virtio-devices/src/watchdog.rs:160]: I suppose this comment is a copy-paste from virtio-rng?

File: virtio-devices/src/watchdog.rs
Comment (sameo) [reply virtio-devices/src/watchdog.rs:42]: I agree this should eventually be added, the driver should be able to fetch the device's default timer interval as opposed to having an implicit assumption about it.
And yes, this can be done as a second step.

File: virtio-devices/src/watchdog.rs
Comment (sameo) [inline virtio-devices/src/watchdog.rs:55]: One suggestion: Instead of timestamping the last ping, we could initially set the timer up and then resetting it (disarm,arm) when receiving a ping. A timer expiration would mean we need to reset, there would no longer need to track the ping time and computed elapsed times since last ping in the timer expiration handler.

File: virtio-devices/src/watchdog.rs
Comment (rbradford) [reply virtio-devices/src/watchdog.rs]: Nice, thanks, fixed.

File: virtio-devices/src/watchdog.rs
Comment (rbradford) [reply virtio-devices/src/watchdog.rs:55]: That is an interesting suggestion thank you @sameo. Currently it takes advantage of the periodic timer so it is not necessary to teardown/setup the timer each time which I feel is a bit cleaenr. It also took a bit of effort to make sure everything worked correctly across pause/resume (and by extension snapshot/restore) so i'm loath to fiddle with the state mechanics now that I know that is working correctly!

**Comment by rbradford** [comment]
I updated this PR based on @sameo's fixes and to use the main branch of https://github.com/cloud-hypervisor/linux

**Review by sboeuf — APPROVED** [review]

**Pull request merged**
